File: 15_Semantic_Rules/ets_cases/15.1_Semantic_Essentials/cross_lang_verify/SEM_15_01_Essentials.java
/**
 * Java cross-language verification for 15.1 Semantic Essentials
 *
 * Key finding: ArkTS semantic essentials are highly consistent with Java.
 * - Numeric widening (int→double) same
 * - String concatenation with + same
 * - Assignment context rules similar
 */
class SEM_15_01_Essentials {
    static int pass = 0, fail = 0;
    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    public static void main(String[] args) {
        // Numeric widening: int→double
        int i = 42;
        double d = i;
        check(d == 42.0, "numeric widening int→double (Java: same)");

        // String concatenation
        String s = "Hello, " + 42 + true;
        check(s.equals("Hello, 42true"), "string concat (Java: same)");

        // Assignment context: subtype
        Object o = "string";
        check(o instanceof String, "assignment context: subtype (Java: same)");

        // Standalone expression type
        int expr = 1 + 2;
        check(expr == 3, "standalone expression int (Java: same)");

        // Variable initialization
        String v = "hello";
        check(v.equals("hello"), "variable initialization (Java: same)");

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) System.exit(1);
        System.out.println("verified");
    }
}

File: 15_Semantic_Rules/ets_cases/15.1_Semantic_Essentials/cross_lang_verify/SEM_15_01_Essentials.swift
/**
 * Swift cross-language verification for 15.1 Semantic Essentials
 *
 * Key difference: Swift requires explicit numeric conversion.
 * ArkTS/Java allow implicit int→double widening.
 */
import Foundation

var pass = 0, fail = 0
func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

// Numeric widening: Swift requires explicit conversion
let i: Int = 42
// let d: Double = i  // COMPILE ERROR in Swift
let d = Double(i)
check(d == 42.0, "numeric widening (Swift: explicit conversion required)")

// String concatenation
let s = "Hello, \(42)\(true)"
check(s == "Hello, 42true", "string interpolation (Swift: pattern)")

// Assignment context: subtype
let o: Any = "string"
check(o is String, "assignment context: subtype (Swift: same)")

// Standalone expression type
let expr = 1 + 2
check(expr == 3, "standalone expression Int (Swift: same)")

// Variable initialization
let v = "hello"
check(v == "hello", "variable initialization (Swift: same)")

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")
